src/librustc_mir/borrow_check.rs

```diff
@@ -1248,13 +1273,356 @@ enum NoMovePathFound {
     ReachedStatic,
 }
 
+/// The degree of overlap between 2 places for borrow-checking.
+enum Overlap {
+    /// The places might partially overlap - in this case, we give
+    /// up and say that they might conflict. This occurs when
+    /// different fields of a union are borrowed. For example,
+    /// if `u` is a union, we have no way of telling how disjoint
+    /// `u.a.x` and `a.b.y` are.
+    Arbitrary,
+    /// The places are either completely disjoint or equal - this
+    /// is the "base case" on which we recur for extensions of
+    /// the place.
+    EqualOrDisjoint,
+    /// The places are disjoint, so we know all extensions of them
+    /// will also be disjoint.
+    Disjoint,
+}
+
 impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+    // Given that the bases of `elem1` and `elem2` are always either equal
+    // or disjoint (and have the same type!), return the overlap situation
```

Comment:
does "equal or disjoint" mean "not known to be disjoint", basically? I guess it is referencing the `EqualOrDisjoint` variant? The phrasing as is sounds a bit like something that can't be false (i.e., sort of like "given that X is either true or false"), though I understand what you mean by it. Maybe just "given that the basis of `elem1` and `elem2` thus far overlap with `Overlap::EqualOrDisjoint`" or something.